### ISSUE TITLE
getting Context for ArrayType not working anymore

### ISSUE THREAD

**Issue by mlangkabel** [javaparser/javasymbolsolver#116]
It seems like in the new version of the Javaparser (including 3.0.0 Alpha 6) the ArrayType is not a real part of the AST anymore. So when the JavaparserFactory asks an ArrayType node for its parent in order to provide a Context, the parent is just a null value.

**Comment by mlangkabel** [comment]
Oh, maybe this is actually a bug in the Javaparser, as the SymbolSolver asks some type node (lets say a ClassOrInterfaceType) for its parent which returns the ArrayType node (that is not an actual part of the AST).
Right now I cannot test if that has already been fixed in the Javaparser because the SymbolSolver is not up to date with the latest Javaparser release.

**Comment by ftomassetti** [comment]
Oh... that is an interesting case... We will need to find the context somehow because otherwise we cannot understand how to solve symbols there. I will first update to the newest version and then check this bug.

**Comment by mlangkabel** [comment]
Awesome! I just created a tiny example code to reproduce this issue:

```
public class JavaTest {
	public String[] field;

	public void foo(String[] arg){
	}
}
```

Everything seems to be working fine in the case of the field declaration but we run into trouble when trying to solve the type of `foo`'s argument.

**Comment by matozoid** [comment]
You're actually right that the type of arg is not part of the AST. It's put together on demand by the getter. You can always give it a parent if you like, but beware of side effects.

I can see why @ftomassetti wants to get rid of all the "bracket" methods :)

**Comment by ftomassetti** [comment]
After migrating to the last version of JP I tried with the example given by @mlangkabel (I just added `arg` in the body of the foo method). 

This test pass:
```
    @Test
    public void arrayTypeIsNotPartOfTheTree() throws ParseException {
        CompilationUnit cu = parseSample("Issue116");
        ClassOrInterfaceDeclaration clazz = Navigator.demandClass(cu, "JavaTest");
        MethodDeclaration methodDeclaration = Navigator.demandMethod(clazz, "foo");
        TypeSolver typeSolver = new ReflectionTypeSolver();
        JavaParserFacade javaParserFacade = JavaParserFacade.get(typeSolver);
        Type type = javaParserFacade.convert(methodDeclaration.getParameters().get(0).getType(), methodDeclaration);
        assertEquals("java.lang.String[]", type.describe());

        ExpressionStmt expressionStmt = (ExpressionStmt) methodDeclaration.getBody().getStmts().get(0);
        Expression argRef = expressionStmt.getExpression();
        assertEquals("java.lang.String[]", javaParserFacade.getType(argRef).describe());
    }
```

But yes, the parent of ArrayType is null. However I cannot find a situation which causes JavaSymbolSolver to crash or return a wrong value. It is not a trivial problem to solve. Maybe when creating the context we could manually set some parent.

@mlangkabel let's see if we can find counter-examples that shows we still have problems caused by this

**Comment by ftomassetti** [comment]
Test added via 15bdafd38d686a6455df4fe07aef11119f308245

**Comment by mlangkabel** [comment]
@ftomassetti thanks for acting so quickly! In the context I'm using the JavaSymbolSolver in I often find myself calling the `javaParserFacade.convert` method handing the same argument over as both of the methods parameters. So I'm calling convert(typeNode, typeNode) because in my dynamic environment I don't always know which node would be best for starting to look for a suitable context.

I think that test should fail if you'd call the convert method like that.

**Comment by mlangkabel** [comment]
Ok I just had some time to validate that. The test fails if we change it to this:
```
        ...
        com.github.javaparser.ast.type.Type<?> typeNode = methodDeclaration.getParameters().get(0).getType();
        Type type = javaParserFacade.convert(typeNode, typeNode);
        ...
``` 
For this case the `convert` method throws a null pointer exception.

**Comment by matozoid** [comment]
Cool tip for everyone: if you quote code, you can turn on syntax highlighting by typing "quote quote quote java" as the opener. So just the word java directly against the quotes.

**Comment by ftomassetti** [comment]
@mlangkabel I think that there is not much we can do on the JSS side: typeNode does not know about the rest of the AST. We could invoke `convert(typeNode, methodDeclaration)` and that should work but `convert(typeNode, typeNode)` cannot not possibly work. If we want it to work we would need to do some changes on the JP side

**Comment by mlangkabel** [comment]
Ok, I think it may be a bug in the Javaparser, so I opened an issue with all the details over there (https://github.com/javaparser/javaparser/issues/574) to clear things up.

**Comment by mlangkabel** [comment]
Updating Javaparser to the `alpha.10` tag fixed this issue. Is it alright if I update the test to use `convert(typeNode, typeNode)`?

**Comment by ftomassetti** [comment]
sure!

**Comment by mlangkabel** [comment]
oh, I think I posted this too early. It seems I was using a quite old version of the symbolsolver to evaluate this. Nevertheless, @matozoid mentioned that the issue should be fixed by now, so we should update the javasymbolsolver to the  new tag ;)

**Comment by ftomassetti** [comment]
just migrated to use alpha.10 so we can check if we have the problem again

**Comment by mlangkabel** [comment]
Awesome! You are the best, thanks :) I'll give it a try first thing tomorrow morning.

**Comment by mlangkabel** [comment]
It's still not working correctly but I finally managed to track down that issue in the JavaParser codebase. I think we will need one more migration once that issue is fixed.

**Comment by ftomassetti** [comment]
thanks for all your work on this one!

**Comment by matozoid** [comment]
This should work correctly again after https://github.com/javaparser/javaparser/pull/615 !

**Comment by ftomassetti** [comment]
Good, we will migrate to 3.0.0 RC.3 as soon as it is released

**Comment by mlangkabel** [comment]
fully fixed with merge of pull request #126
